addon/components/tri-state.js

```diff
@@ -30,60 +32,77 @@ export default Component.extend({
   layout,
 
   /**
-   * A service that is event aware so we can trigger actions in tri-state from external sources.
-   * Note this is NOT RECOMMENDED as it goes against DDAU and has the potential to introduce memory
-   * leaks if we don't clean up after ourselves. However, it can be useful in certain situations.
-   * @type {Object}
+   * Cache for the last resolved data to be shown during requests when showLastSuccessful is true
+   * @type {*}
+   * @private
    */
-  triEvents: inject.service(),
+  _lastResolvedData: null,
 
   /**
-   * Flag used to indicate if we should wait for all promises to resolve or
-   * if we should reject immediately as soon as one fails.
-   * @type {Boolean}
+   * Helper function to resolve promises based on how they are provided
+   * @param {Promise|Array|Object} promises The promises to resolve
+   * @return {Promise} Promise which resolves when all promises have resolved
+   * @private
    */
-  resolveAll: computed.not('failFast'),
+  _resolvePromises(promises) {
+    if (promises instanceof Promise) {
+      return promises;
+    }
+
+    if (Array.isArray(promises)) {
+      return RSVP.all(promises);
+    }
+
+    if (typeof promises === 'object') {
+      return RSVP.hash(promises);
+    }
+
+    return promises;
+  },
 
   /**
-   * Task aliases
+   * Computes promise state after all promises are either resolved or rejected
+   * @return {Array} Array containing the promise state and data associated with the promise state
+   * @private
    */
-  isLoading: computed.alias('taskInstance.isRunning'),
-  isError: computed.bool('taskInstance.error'),
-  isSuccess: computed.bool('taskInstance.value'),
-  lastSuccessValue: computed.alias('_fetchDataTask.lastSuccessful.value'),
-  hasSuccessData: computed.bool('lastSuccessValue'),
+  _computePromiseState() {
```

Comment:
Let's get state via a computed property that wraps the promises in a PromiseProxyMixin object. That way we get pending, settled, rejected, and fulfilled for free. If we did this we would have to attempt to resolve the promise(s) in a separate method. I'm leaning toward this approach: http://blog.deveo.com/promiseproxymixin-pure-ember-alternative-to-ember-concurrency/